### ISSUE TITLE
LLVM subprojects that use their own tablegen cannot be built with LLVM_ENABLE_WARNINGS=OFF

### ISSUE THREAD

**Issue by RIscRIpt** [llvm/llvm-project#112528]
Building `mlir` with `-DLLVM_ENABLE_WARNINGS=OFF` results in an error that `-no-warn-on-unused-template-args` is not defined for internal tablegen executable `mlir-src-sharder`.

```bash
cmake -G "Ninja" -DLLVM_ENABLE_PROJECTS="mlir" -DLLVM_ENABLE_WARNINGS=OFF -B build/Debug llvm
cmake --build build/Debug
```

```
mlir-src-sharder: Unknown command line argument '-no-warn-on-unused-template-args'.  Try: '/Users/riscript/projects/llvm-project/build/Debug/NATIVE/bin/mlir-src-sharder --help'
```

The problem is caused by [`llvm/camke/modules/TableGen.cmake::tablegen`](https://github.com/llvm/llvm-project/commit/d968b173d33be1a5819e9f5f9a8b4258775a65ed#diff-af6c3a2cba6f2d0e48d7e6f6c6ab7084527c8083d5c3ba64f02babec58e0a7eaR84) which adds `-no-warn-on-unused-template-args` to any tablegen executable.

```cmake
  if (NOT LLVM_ENABLE_WARNINGS)
    list(APPEND LLVM_TABLEGEN_FLAGS "-no-warn-on-unused-template-args")
  endif()
```

However, some other subprojects (e.g. mlir) can use CMake function `tablegen` from LLVM with their own tablegen executables that do not have this flag, e.g. [`add_sharded_ops`](https://github.com/llvm/llvm-project/commit/1b232fa0e9864dde230db8da82a906c588baf792#diff-ac43f9356d1ba6701fbc1dcf42ebce1250eec9db8bc5d1d539738e89b8b8ab4bR183).

Currently, to build LLVM one should either avoid using `-DLLVM_ENABLE_WARNINGS=OFF`, or set `-DLLVM_INCLUDE_TESTS=OFF` (this excludes invocations to `add_sharded_ops`).

### Possible solutions
1. Somehow make `llvm/camke/modules/TableGen.cmake::tablegen` detect non-llvm tablegen and omit `-no-warn-on-unused-template-args` flag
2. Generalize fix from (#55716) 983cb6c92fec7b39ec9760f1410d3003484df3a9 to other tablegen usages in MLIR.
